### PR TITLE
Introduce dedicated error types to the lapin-futures crate

### PR DESCRIPTION
**Pull request by kureuil**
This change replaces all occurrences of `io::Error` by the newly created
`Error` type, the only exception being for `AMQPCodec`. This change allows
future-proofing the error that can be returned by the crate. For example
by changing the internals of the `Error` type, it is now possible to
provide users with a way of knowing what kind of error occurred without
having to resort to parsing error messages.

The only exception to this change concerns `AMQPCodec`: a new `CodecError` type
was created just for it because `tokio-codec`'s `Encoder` and `Decoder` traits
impose a `From<std::io::Error>` bound of the error type. Because I didn't
want to introduce a way to create an `Error` instance from outside the
crate, I isolated the possible errors from this section into its own
type.

### CODE REVIEW

File: futures/src/channel.rs [hunk 1]
```diff
@@ -270,7 +270,7 @@ impl<T: AsyncRead+AsyncWrite+Send+Sync+'static> Channel<T> {
                     transport: channel_transport.clone(),
                 }))
             } else {
-                return Err(io::Error::new(io::ErrorKind::ConnectionAborted, "The maximum number of channels for this connection has been reached"));
+                return Err(Error::new("The maximum number of channels for this connection has been reached"));
```
Comment (Keruspe) [inline futures/src/channel.rs]: Maybe we should keep some kind of `io::Error` for actual io errors?

File: futures/examples/client.rs [hunk 2]
```diff
@@ -82,7 +84,7 @@ fn main() {
             c.basic_ack(message.delivery_tag, false)
           })
         })
-      })
+      }).map_err(Error::from)
     }).map_err(|err| eprintln!("error: {:?}", err))
```
Comment (Keruspe) [inline futures/examples/client.rs]: From what I understand, we could/should drop the `:?` in places like that, right?

File: futures/src/transport.rs [hunk 3]
```diff
@@ -30,15 +48,15 @@ pub struct AMQPCodec {
...
 
-    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<AMQPFrame>, io::Error> {
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<AMQPFrame>, CodecError> {
         let (consumed, f) = match parse_frame(buf) {
           Err(e) => {
             if e.is_incomplete() {
               return Ok(None);
             } else {
-              return Err(io::Error::new(io::ErrorKind::Other, format!("parse error: {:?}", e)));
+              return Err(CodecError::ParseError);
```
Comment (Keruspe) [inline futures/src/transport.rs]: Don't we lose the error there?

File: futures/examples/topic.rs [hunk 4]
```diff
@@ -21,27 +21,27 @@ fn main() {
     let runtime = Runtime::new().unwrap();
 
     runtime.block_on_all(
-        TcpStream::connect(&addr).and_then(|stream| {
+        TcpStream::connect(&addr).map_err(Error::from).and_then(|stream| {
```
[the remark this answers is not in the document]
Comment (kureuil) [reply futures/examples/topic.rs]: Also note that once async/await lands in stable, we'll be able to use the `?` operator which automatically performs these conversions.

**Comment by kureuil** [comment]
@Keruspe rebased onto current master, awaiting for your review :)

File: futures/src/channel.rs (hunk 1 above)
Comment (Keruspe) [reply futures/src/channel.rs]: Right, fine by me then !

**Comment by Keruspe** [comment]
Thanks a lot for this. Always wanted to give failure a look and never had the occasion before

**Pull request merged**